pyamg/util/utils.py

```diff
@@ -2142,28 +2142,68 @@ def truncate_rows(A, nz_per_row):
     return A
 
 
-# from functools import partial, update_wrapper
-# def dispatcher(name_to_handle):
-#    def dispatcher(arg):
-#        if isinstance(arg,tuple):
-#            fn,opts = arg[0],arg[1]
-#        else:
-#            fn,opts = arg,{}
-#
-#        if fn in name_to_handle:
-#            # convert string into function handle
-#            fn = name_to_handle[fn]
-#        #elif isinstance(fn, type(numpy.ones)):
-#        #    pass
-#        elif callable(fn):
-#            # if fn is itself a function handle
-#            pass
-#        else:
-#            raise TypeError('Expected function')
-#
-#        wrapped = partial(fn, **opts)
-#        update_wrapper(wrapped, fn)
-#
-#        return wrapped
-#
-#    return dispatcher
+def scale_block_inverse(A, blocksize):
+    """
+    Get inverse of block diagonal of A and scale A, A = D^{-1}A.
+
+    Parameters
+    ----------
+    A : csr or bsr_matrix
+        Matrix to scale by block inverse. 
+    blocksize : int
+        Blocksize of matrix.
+
+    Returns
+    -------
+    tuple, (D^{-1}*A, D^{-1})
+
+    Notes
+    -----
+    This is not a symmetric scaling. 
+
+    Examples
+    -----
+    >>> from pyamg.gallery import poisson
+    >>> from pyamg.util.utils import scale_block_inverse
+    >>> A = poisson((4,), format='csr')
+    >>> A, Dinv = scale_block_inverse(A, 2)
+    >>> A.toarray()
+    array([[ 1.        ,  0.        , -0.33333333,  0.        ],
+           [ 0.        ,  1.        , -0.66666667,  0.        ],
+           [ 0.        , -0.66666667,  1.        ,  0.        ],
+           [ 0.        , -0.33333333,  0.        ,  1.        ]])
+    """
+
+    if not isspmatrix(A):
+        raise TypeError('Expected sparse matrix')
+    if A.shape[0] != A.shape[1]:
+        raise ValueError("Expected square matrix")
+    if np.mod(A.shape[0], blocksize) != 0:
+        raise ValueError("blocksize and A.shape must be compatible")
+
+    # Convert to BSR
+    if not isspmatrix_bsr(A):
+        A = bsr_matrix(A, blocksize=(blocksize, blocksize))
+    if A.blocksize != (blocksize, blocksize):
+        A = A.tobsr(blocksize=(blocksize, blocksize))
+
+    # Get block diagonal inverse
+    N_block = A.shape[0] / blocksize
+    Dinv = get_block_diag(A=A, blocksize=blocksize, inv_flag=True)
+    scale = bsr_matrix((Dinv, np.arange(0,N_block), np.arange(0,N_block+1)),
+                        blocksize=[blocksize,blocksize], shape=A.shape)
+    return scale * A, scale
+
+
+# Get the square diagonal blocks of csr_matrix A
```

Comment:
I don't think `extract_diagonal_blocks` is used anywhere.  Can we delete it?  

If not, then it will need a unit test (just something real simple like for scale_block_inverse) and a docstring

Reply:
Ya I'm not seeing it anywhere. The goal I can think of is to be able to take a system Ax=b, scale both sides by block diagonal inverse, then apply AMG to the scaled system, and scale_block_inverse allows this; not sure what the extract_diagonal_blocks woudl be needed for; it just seems to allow variable block sizes, but nothing else in pyamg does.. will think on it, but can probably remove

Reply:
Pylint is complaining about it, so I'm going to remove it for now.  We can always add it back later if needed